File: OverlayManager/TS/Game Engine/TargetValue.ts
enum TargetBinding {
  truncate,
  wrap,
  rock
}

class TargetValue {
  absoluteVariance: number;
  drift: number;
  binding: TargetBinding = TargetBinding.truncate;
  constructor(public target: number, public relativeVariance: number = 0, public lowBounds: number = 0, public highBounds: number = Infinity) {

  }

  getValue(): number {
    var result: number;

    if (this.absoluteVariance != undefined)
      result = Random.getVarianceAbsolute(this.target, this.absoluteVariance);
    else
      result = Random.getVarianceRelative(this.target, this.relativeVariance);

    if (this.drift != undefined) {
      this.target += this.drift;
    }

    if (this.binding === TargetBinding.truncate)
      return MathEx.truncate(result, this.lowBounds, this.highBounds);
    else if (this.binding === TargetBinding.wrap)
      return MathEx.wrap(result, this.lowBounds, this.highBounds);
    else if (this.drift != undefined) // rock
    {
      if (result < this.lowBounds) {
        result = this.lowBounds;
        this.drift = Math.abs(this.drift);
      }
      else if (result > this.highBounds) {
        result = this.highBounds;
        this.drift = -Math.abs(this.drift);
      }
      return result;
    }
  }
}

File: DHDM/DnD Model/Effects/TargetValue.cs
namespace DHDM
{
	public class TargetValue
	{
		public double value;
		public double relativeVariance;
		public double absoluteVariance;
		public double min;
		public double max;
		public double drift;
		public TargetBinding targetBinding;
		public TargetValue(double value, double relativeVariance = 0, double absoluteVariance = 0, double min = 0, double max = 1, double drift = 0, TargetBinding targetBinding = TargetBinding.truncate)
		{
			this.value = value;
			this.relativeVariance = relativeVariance;
			this.absoluteVariance = absoluteVariance;
			this.min = min;
			this.max = max;
			this.drift = drift;
			this.targetBinding = targetBinding;
		}

		public TargetValue()
		{
		}
	}
}
